snekbox/memfs.py

```diff
@@ -0,0 +1,137 @@
+    )
+    return tmp
+
+
+def unmount_tmpfs(name: str) -> None:
+    """Unmount and remove a tmpfs directory."""
+    tmp = NAMESPACE_DIR / name
+    subprocess.check_call(["umount", str(tmp)])
+    # Unmounting will not remove the original folder, so do that here
+    tmp.rmdir()
+
+
+class MemFS:
+    """A temporary directory using tmpfs."""
+
+    assignment_lock = BoundedSemaphore(1)
+    assigned_names: set[str] = set()  # Pool of tempdir names in use
+
+    def __init__(self, instance_size: int) -> None:
+        """
+        Create a temporary directory using tmpfs.
+
+        size: Size limit of each tmpfs instance in bytes
+        """
+        self.path: Path | None = None
+        self.instance_size = instance_size
+
+    @property
+    def name(self) -> str | None:
+        """Name of the temp dir."""
+        return self.path.name if self.path else None
+
+    @property
+    def home(self) -> Path | None:
+        """Path to home directory."""
+        return Path(self.path, "home") if self.path else None
+
+    def __enter__(self) -> MemFS:
+        """Mounts a new tempfs, returns self."""
+        with self.assignment_lock:
+            for _ in range(10):
+                # Combine PID to avoid collisions with multiple snekbox processes
+                if (name := f"{PID}-{uuid4()}") not in self.assigned_names:
+                    self.path = mount_tmpfs(name, self.instance_size)
+                    self.assigned_names.add(name)
+                    break
+            else:
+                raise RuntimeError("Failed to generate a unique tempdir name in 10 attempts")
+
+        self.mkdir("home")
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        self.cleanup()
+
+    def mkdir(self, path: Path | str, chmod: int = 0o777) -> Path:
+        """Create a directory in the tempdir."""
+        folder = Path(self.path, path)
+        folder.mkdir(parents=True, exist_ok=True)
+        folder.chmod(chmod)
+        return folder
+
+    def attachments(
+        self, max_count: int, max_size: int | None = None
+    ) -> Generator[FileAttachment, None, None]:
+        """Return a list of attachments in the tempdir."""
+        count = 0
+        # Look for any file starting with `output`
+        for file in self.home.glob("output*"):
+            if count > max_count:
+                log.info(f"Max attachments {max_count} reached, skipping remaining files")
+                break
+            if file.is_file():
+                count += 1
+                yield FileAttachment.from_path(file, max_size)
+
+    def cleanup(self) -> None:
+        """Unmounts tmpfs, releases name."""
+        if self.path is None:
+            return
+        # Remove the path folder
+        unmount_tmpfs(self.name)
+
+        if not self.path.exists():
+            with self.assignment_lock:
```

Comment:
Why is a lock necessary?

Reply:
I was thinking there might be issues if we had multiple threads? Though I suppose we don't have that currently. I don't think it hurts to make it thread safe for the time being though?

Reply:
Gunicorn is configured with multiple workers, though right now I cannot recall if those are threads or processes. My vote is for simplicity, so a lock shouldn't be included if it's not necessary. Can you investigate further? If it is needed, then there should also be a test for whatever bad situation this lock prevents.

Reply:
It seems the `path.mkdir()` in `mount_tmpfs` should already be sufficient as the python unix implementation appears to be free from race conditions.

Removed set and semaphore in [8c897dc](https://github.com/python-discord/snekbox/pull/159/commits/8c897dc0e86f4c6cb5e09e376c6aa40d4093ed7d) in favor of catching the `FileExistsError` instead.

Added tests in [b74b813](https://github.com/python-discord/snekbox/pull/159/commits/b74b813ae6ac2ebf374cc12bb0e9689f49776f2e) to ensure this new approach works fine in multithread conditions.